### ISSUE TITLE
Bug

### ISSUE THREAD

**Issue by ElPichiii** [markusfisch/PieLauncher#119]
Sometimes the Launcher doesnt work, you what yo open WhatsApp (for example) and the Launcher doesnt show the circle with the apps, and open always the same app

**Comment by ElPichiii** [comment]
yeah, is not always but when that happen is annoying, yes, i have in the hold menu whatsapp, for example, and when i hold, nothing appears and opens spotify all the time, and when i tap, apps doesnt appear and i have to write the name

**Comment by devilsclaw** [comment]
I am not using whatsapp but in general when I use the 3 button soft keys circle button to return to the desktop. At times when I push and hold my finger on the screen it wont show. I have also seen where when its in this state that if I press a combination of the square button or the circle it will show the pie launcher draw and sometime is interact able and other times its not. I have rebooted and fixes it for a while and also killing it through the app settings section works also.

**Comment by devilsclaw** [comment]
So this happens multiple times a day for me. Commonly when I wake up the phone and use pattern to unlock.

It seems to be stuck in the state where it thinks that it's in the draw, but it shows the desktop and not the icon. 

Other times it's as if it's stuck with the wheel open but. No icons showing.

This will also happen when hopping between apps.

**Comment by ElPichiii** [comment]
For me is sometimes when i go back to the main screen and sometimes when the phone is blocked and i use my finger to unblock it, and i dont think they draw is stuck or invisible, cuz when this happen to me and i try to open other app, it always open the same one

**Comment by devilsclaw** [comment]
Broken states:
1) wheel is invisible but last action will always be repeated
2) Drawer is invisible but can select icons. Can even bring up the keyboard

With a combination of the three soft key actions will cause state 2 if state 1 is already happening. I happens more often after the phone comes out of a prolonged sleep. But it can happen because of context twitching between apps.

**Comment by markusfisch** [comment]
Thanks for these observations! It's really hard to tell what's going on here since I never seem to get into these broken states 🙈 

> With a combination of the three soft key actions will cause state 2 if state 1 is already happening.

What combination exactly? Is there a reproducible sequence of actions that will lead to one these situations?

**Comment by devilsclaw** [comment]
I don't know what gets in state 1 but if you press the circle while it's in that state it will bring up the drawer but it has two states also sometime it will be stuck on the screen visibly but most of the time it's invisible.

**Comment by kinetikeith** [comment]
> Broken states:
> 
> 1. wheel is invisible but last action will always be repeated
> 2. Drawer is invisible but can select icons. Can even bring up the keyboard
> 
> With a combination of the three soft key actions will cause state 2 if state 1 is already happening. It happens more often after the phone comes out of a prolonged sleep. But it can happen because of context switching between apps.

This happens to me as well on about a daily basis, a reboot always fixes it. Also, I've found that clearing all recent apps when the (invisible) app drawer is open seems to fix it sometimes.

**Comment by devilsclaw** [comment]
So I fixed the problem for me. but its not a real solution for the problem but it might point to the problem location.

I moved to have the draw showing up only if I select it in the menu via an icon. So tap or touch to draw is now in essence disabled.

**Comment by markusfisch** [comment]
> I moved to having the draw showing up only if I select it in the menu via an icon. So tap or touch to draw is now in essence disabled.

Sorry, I don't quite understand 🙈  Did you change the code (and if so, what exactly)?

**Comment by devilsclaw** [comment]
No. I just changed the option in how to get to the draw. from Opens with "Tap on home screen (Default)" to "Icon in pie menu". After that I no longer have the problem. I think there must be false positives happening or a stuck state related to the Tap on home screen option.

**Comment by markusfisch** [comment]
Ah, thanks for the clarification 👍

**Comment by devilsclaw** [comment]
Well the problem is reduced greatly but I have now had it happen twice in the last couple days. Still much reduced. Each time has been coming out of sleep now.

**Comment by devilsclaw** [comment]
I figured out how to cause it. In my. Case I have fennec fox browser left in the forefront / on screen. Then I was 30 seconds to a minute. Turn my phone back on. Enter a pattern lock screen to unlock. The once logged in I press the square of the three key navigation. Then I press the circle of the three button navigation. Then tap or old anywhere on the desktop and it will be broken.

**Comment by devilsclaw** [comment]
I'm going to have to add. That my phone is a pixel fold. And it happens when everything above happens when going from folded to unfolded and the doing what is listed above. If I do it on the front screen in the folded state it does not happen.

**Comment by devilsclaw** [comment]
I got it to happen on front screen but only after the waiting for 30 to 60 seconds. Then unfold the phone . Then fold the phone again. Then do the login and everything else.

**Comment by devilsclaw** [comment]
Rotating does not cause it. So its screen size change. That causes it.

**Comment by devilsclaw** [comment]
So one more thing to add. If during the unfold part if I switch from landscape to portrait orientation, before or after login the problem also won't happen.

**Comment by markusfisch** [comment]
Good catch! 👍 Thank you very much for these observations! Really appreciated!

**Comment by devilsclaw** [comment]
I found one more today. When its in the broken state. If I cause the window to rotate from landscape to portrait it will also recover from the problem.

**Comment by devilsclaw** [comment]
Well here is another very interesting thing. I compiled this project my self and the problem does not exist on my build. I renamed it to Pie Luancher2 and renamed the correct directory so I can run both side by side.

Then I setup the second instance the same as the original.

I am  using the android sdk 34, also I have used both openjdk 17 and openjdk 21 for the job. both don't show the problem I have been experiencing. I thought the problem might of been due to the java version and maybe you are building with an even newer version that 21?

**Comment by devilsclaw** [comment]
Well it finally happened today on my build and even on the front screen. This is starting to feel like a race condition or a unprotected variable and multi threading.
